File: cpp_url_fetch/src/value_parser.cpp
#include <cstdint>
#include <string>
#include <optional>
#include <vector>

#include "value_parser.hpp"

std::string to_string(const std::vector<uint8_t> & vec) {
    std::string out;
    out.reserve(vec.size());

    for (uint8_t b : vec) {
        out += b;
    }

    return out;
}

// "bytes" array must be exactly 4 bytes long!
uint32_t from_bigend_bytes_32(const uint8_t * const bytes) {
    return (uint32_t)(bytes[0]) << 24
        | (uint32_t)(bytes[1]) << 16
        | (uint32_t)(bytes[2]) << 8
        | (uint32_t)(bytes[3]);
}

bool deserialize_value(Value & result, const std::vector<uint8_t> & source_bytes) {
    std::string url;
    std::optional<std::string> password;

    enum class Member {
        Url,
        Password,
        None
    };
    enum class Field {
        Size,
        Data
    };

    Member current_member = Member::Url;
    Field current_field = Field::Size;

    uint8_t size_buf[4];
    int size_buf_i = 0;
    uint32_t size;

    std::vector<uint8_t> data;

    for (const uint8_t byte : source_bytes) {
        if (current_member == Member::None) {
            return false;
        }

        switch (current_field) {
            case Field::Size: {
                size_buf[size_buf_i++] = byte;
                if (size_buf_i >= 4) {
                    size = from_bigend_bytes_32(size_buf);
                    size_buf_i = 0;
                    current_field = Field::Data;
                }
                break;
            }
            case Field::Data: {
                if (size <= 0) {
                    // This should not happen
                    return false;
                }
                data.push_back(byte);
                size--;
                break;
            }
            default: {
                return false;
            }
        }

        if (current_field == Field::Data) {
            if (size <= 0) {
                switch (current_member) {
                    case Member::Url:
                        url = to_string(data);
                        current_member = Member::Password;
                        current_field = Field::Size;
                        data.clear();
                        break;
                    case Member::Password:
                        password = to_string(data);
                        current_member = Member::None;
                        current_field = Field::Size;
                        data.clear();
                        break;
                    default:
                        // This should not happen
                        return false;
                }
            } else if (size == UINT32_MAX) {
                switch (current_member) {
                    case Member::Password:
                        password = std::nullopt;
                        current_member = Member::None;
                        current_field = Field::Size;
                        break;
                    default:
                        // continue normally
                        break;
                }
            }
        }
    }

    if (current_member != Member::None) {
        return false;
    }

    result = {url, password};
    return true;
}

File: rust_url_fetch/src/value_parser.rs
// Value that will be stored in the KV store.
pub struct Value {
    pub url: String,
    pub password: Option<String>,
}

// Error returned when deserialization of the Value fails.
pub enum DeserializeError {
    TooShort,
    TooLong,
    InternalError,
}

// Reconstruct Value from the serialized bytes, which look like this:
// | URL_LENGTH | URL DATA | PASSWORD_LENGTH | PASSWORD_DATA |
pub fn deserialize_value(bytes: &Vec<u8>) -> Result<Value, DeserializeError> {
    let mut url: String = String::new();
    let mut password: Option<String> = None;

    enum Member {
        Url,
        Password,
        None,
    }
    enum Field {
        Size,
        Data,
    }

    let mut current_member = Member::Url;
    let mut current_field = Field::Size;

    let mut size_buf: [u8; 4] = [0; 4];
    let mut size_buf_i = 0;
    let mut size: u32 = 0;

    let mut data: Vec<u8> = Vec::new();

    for byte in bytes {
        if matches!(current_member, Member::None) {
            return Err(DeserializeError::TooLong);
        }

        match current_field {
            Field::Size => {
                size_buf[size_buf_i] = *byte;
                size_buf_i += 1;
                if size_buf_i >= 4 {
                    size = u32::from_be_bytes(size_buf);
                    size_buf_i = 0;
                    current_field = Field::Data;
                }
            }
            Field::Data => {
                if size <= 0 {
                    // This should not happen
                    return Err(DeserializeError::InternalError);
                }
                data.push(*byte);
                size -= 1;
            }
        }

        if matches!(current_field, Field::Data) {
            if size <= 0 {
                match current_member {
                    Member::Url => {
                        url = String::from_utf8(data).unwrap();
                        current_member = Member::Password;
                        current_field = Field::Size;
                        data = Vec::new();
                    }
                    Member::Password => {
                        password = Some(String::from_utf8(data).unwrap());
                        current_member = Member::None;
                        current_field = Field::Size;
                        data = Vec::new();
                    }
                    _ => {
                        // This should not happen
                        return Err(DeserializeError::InternalError);
                    }
                }
            } else if size == u32::MAX {
                match current_member {
                    Member::Password => {
                        password = None;
                        current_member = Member::None;
                        current_field = Field::Size;
                    }
                    _ => {
                        // continue normally
                    }
                }
            }
        }
    }

    if !matches!(current_member, Member::None) {
        return Err(DeserializeError::TooShort);
    }

    return Ok(Value { url, password });
}
